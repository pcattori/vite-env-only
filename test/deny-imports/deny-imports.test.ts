import path from "node:path"

import * as vite from "vite"
import { describe, expect, test } from "vitest"

import denyImports from "../../src/deny-imports"

const root = __dirname

const config = ({
  ssr,
  options,
}: {
  ssr: boolean
  options: Parameters<typeof denyImports>[0]
}): vite.InlineConfig => ({
  root: __dirname,
  build: {
    ssr,
    minify: false,
    lib: {
      entry: "lib/main.js",
      formats: ["es"],
    },
    rollupOptions: {
      output: {
        dir: ssr
          ? path.join(root, "dist/server")
          : path.join(root, "dist/client"),
        entryFileNames: "index.js",
      },
    },
  },
  plugins: [denyImports(options)],
})

describe("denyImports", () => {
  for (let env of ["server", "client"]) {
    let ssr = env === "server"
    let otherEnv = env === "server" ? "client" : "server"

    test(`specifiers / denied by exact match [${env}]`, async () => {
      await expect(
        vite.build(
          config({
            ssr,
            options: {
              [env]: { specifiers: ["./deny-me"] },
            },
          })
        )
      ).rejects.toMatchSnapshot()
    })

    test(`specifiers / denied by regex [${env}]`, async () => {
      await expect(
        vite.build(
          config({
            ssr,
            options: {
              [env]: { specifiers: [/^\.\/deny-me$/] },
            },
          })
        )
      ).rejects.toMatchSnapshot()
    })

    test(`specifiers / ignores other envs [${env}]`, async () => {
      vite.build(
        config({
          ssr,
          options: {
            [otherEnv]: { specifiers: [/^\.\/deny-me\.js$/] },
          },
        })
      )
    })

    test(`specifiers / passes [${env}]`, async () => {
      vite.build(
        config({
          ssr,
          options: {
            [env]: { specifiers: ["other", "**/other/*", /other/] },
          },
        })
      )
    })

    test(`files / denied by exact match [${env}]`, async () => {
      await expect(
        vite.build(
          config({
            ssr,
            options: {
              [env]: { files: ["lib/deny-me.js"] },
            },
          })
        )
      ).rejects.toMatchSnapshot()
    })

    test(`files / denied by glob [${env}]`, async () => {
      await expect(
        vite.build(
          config({
            ssr,
            options: {
              [env]: { files: ["**/deny-me.*"] },
            },
          })
        )
      ).rejects.toMatchSnapshot()
    })

    test(`files / denied by regex [${env}]`, async () => {
      await expect(
        vite.build(
          config({
            ssr,
            options: {
              [env]: { files: [/^lib\/deny-me\.js$/] },
            },
          })
        )
      ).rejects.toMatchSnapshot()
    })

    test(`files / passes [${env}]`, async () => {
      vite.build(
        config({
          ssr,
          options: {
            [env]: { files: ["other", "**/other/*", /other/] },
          },
        })
      )
    })
  }
})
