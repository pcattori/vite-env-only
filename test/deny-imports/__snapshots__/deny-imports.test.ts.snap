// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`denyImports > files / denied by exact match [client] 1`] = `
[Error: [vite-env-only] Import denied
 - Denied by file pattern: lib/deny-me.js
 - Importer: lib/main.js
 - Import: "./deny-me"
 - Resolved: lib/deny-me.js
 - Environment: client]
`;

exports[`denyImports > files / denied by exact match [server] 1`] = `
[Error: [vite-env-only] Import denied
 - Denied by file pattern: lib/deny-me.js
 - Importer: lib/main.js
 - Import: "./deny-me"
 - Resolved: lib/deny-me.js
 - Environment: server]
`;

exports[`denyImports > files / denied by glob [client] 1`] = `
[Error: [vite-env-only] Import denied
 - Denied by file pattern: **/deny-me.*
 - Importer: lib/main.js
 - Import: "./deny-me"
 - Resolved: lib/deny-me.js
 - Environment: client]
`;

exports[`denyImports > files / denied by glob [server] 1`] = `
[Error: [vite-env-only] Import denied
 - Denied by file pattern: **/deny-me.*
 - Importer: lib/main.js
 - Import: "./deny-me"
 - Resolved: lib/deny-me.js
 - Environment: server]
`;

exports[`denyImports > files / denied by regex [client] 1`] = `
[Error: [vite-env-only] Import denied
 - Denied by file pattern: /^lib\\/deny-me\\.js$/
 - Importer: lib/main.js
 - Import: "./deny-me"
 - Resolved: lib/deny-me.js
 - Environment: client]
`;

exports[`denyImports > files / denied by regex [server] 1`] = `
[Error: [vite-env-only] Import denied
 - Denied by file pattern: /^lib\\/deny-me\\.js$/
 - Importer: lib/main.js
 - Import: "./deny-me"
 - Resolved: lib/deny-me.js
 - Environment: server]
`;

exports[`denyImports > specifiers / denied by exact match [client] 1`] = `
[Error: [vite-env-only] Import denied
 - Denied by specifier pattern: ./deny-me
 - Importer: lib/main.js
 - Import: "./deny-me"
 - Environment: client]
`;

exports[`denyImports > specifiers / denied by exact match [server] 1`] = `
[Error: [vite-env-only] Import denied
 - Denied by specifier pattern: ./deny-me
 - Importer: lib/main.js
 - Import: "./deny-me"
 - Environment: server]
`;

exports[`denyImports > specifiers / denied by regex [client] 1`] = `
[Error: [vite-env-only] Import denied
 - Denied by specifier pattern: /^\\.\\/deny-me$/
 - Importer: lib/main.js
 - Import: "./deny-me"
 - Environment: client]
`;

exports[`denyImports > specifiers / denied by regex [server] 1`] = `
[Error: [vite-env-only] Import denied
 - Denied by specifier pattern: /^\\.\\/deny-me$/
 - Importer: lib/main.js
 - Import: "./deny-me"
 - Environment: server]
`;
